src/__tests__/ParseObject-test.js

```diff
@@ -3862,4 +3863,41 @@ describe('ParseObject pin', () => {
     });
     CoreManager.set('ALLOW_CUSTOM_OBJECT_ID', false);
   });
+
+  it('handles unsaved circular references', async () => {
+    const xhrs = [];
+    RESTController._setXHR(function () {
+      const xhr = {
+        setRequestHeader: jest.fn(),
+        open: jest.fn(),
+        send: jest.fn(),
+        status: 200,
+        readyState: 4,
+      };
+      xhrs.push(xhr);
+      return xhr;
+    });
+
+    const a = {};
+    const b = {};
+    a.b = b;
+    b.a = a;
+
+    const object = new ParseObject('Test');
+    object.set('a', a);
+    expect(() => {
+      object.save();
+    }).toThrowError(
+      'Maximum recursive calls exceeded in traverse function. Potential infinite recursion detected.'
+    );
+  });
+
+  it('throws error for infinite recursion', () => {
```

Comment:
This should probably move into a separate file for testing the encode file, as this test is unrelated to the higher level Parse Object?

Reply:
This was [your suggestion](https://github.com/parse-community/Parse-SDK-JS/pull/2099#discussion_r1544635679):

> Doesn't it work to encode the mocked obj with the pointers set? Or maybe it's easier to test on a higher level. Why not just create a test in ParseObject-test.js in which you create a normal Parse Object and set the properties in order to invoke the recursive encode?

I can *try* to create a new separate file if you think it's necessary, but I know it was having issues in the `encode.js` file as we discussed above.

Reply:
Every source file usually has an associated test file. If you test on a higher level ParseObject, then you would add the test to `ParseObject-tes.js`. If you test on a lower level encode, then you'd add the test to `encode-test.js`. You have now added two tests, one using Parse Object, another using encode directly. If you believe we need the encode test in addition to the higher level test, then it should be moved into a separate file. If you believe that they both cover the same, then you could just remove the encode test.